### PR TITLE
Start on removing form cache for ajax.

### LINKED ISSUE #3833: Do not make entries in "cache_form" when viewing forms that use #ajax['callback']
I just ran into an issue on my site where People who are logged out are not able to submit a webform with an ajax component. 

In the error log I see: `Invalid form POST data.`

Drupal issue below

---

### Problem/Motivation
The "cache_form" database table (or cache bin) is well known for being a problem point for heavy-traffic sites or in situations where a form is displayed on every page of the site, such as a login or contact block. The problem is that any form that makes use of the <code>#ajax</code> form API property will immediately make two entries in the <code>cache_form</code> table just by <em>viewing</em> the form.

This has lead to run-away tables with millions of entries in some cases. And while more recent solutions like [#2091511] attempt to mitigate the problem, they don't solve the underlying issue that the Drupal form system has a database <em>write</em> operation simply upon <em>view</em> of a form.

This issue has already been solved in Drupal 8 through [#2263569].

### Proposed resolution
Similar to the Drupal 8 issue, make it so that instead of posting to <code>system/ajax</code>, AJAX requests <em>by default</em> post to the current page path. This allows the normal form execution to occur in the same way as the non-Ajax approach. Then detect when the form is built (in <code>drupal_get_form()</code>) if it was done via an Ajax request. If so, process the form normally, but then return the set of Ajax commands directly, ending the page request early.

This eliminates the need for the <code>system/ajax</code> callback entirely, but it is left in place for backwards-compatibility. This suggested change only affects #ajax usages that use "callback", any usages that leverage "path" will still by default trigger the "cache_form" table to be used.

### API changes
Switches the File module to use #ajax['callback'] instead of #ajax['path'], and in the process eliminates the menu callback at "file/ajax".

---
Related:
* [matching d.o issue](https://www.drupal.org/project/drupal/issues/2819375)
* [PS] Use signatures instead of the form cache https://github.com/backdrop/backdrop-issues/issues/1470
* Consider exposing `form_cache_expiration` and/or providing an option so that it can be cleared along with other caches. https://github.com/backdrop/backdrop-issues/issues/3191
---
PR:

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/3833

### CODE REVIEW

File: core/includes/ajax.inc [hunk 1]
```diff
@@ -382,7 +382,16 @@ function ajax_get_form() {
...
   list($form, $form_state, $form_id, $form_build_id, $commands) = ajax_get_form();
   backdrop_process_form($form['#form_id'], $form, $form_state);
+  return ajax_form_return_commands($form, $form_state, $commands);
+}
 
+/**
+ * Return a set of AJAX commands for updating a form.
+ *
+ * @see ajax_form_callback()
+ * @see drupal_get_form()
```
Comment (klonos) [inline core/includes/ajax.inc:392]:
```suggestion
 * @see backdrop_get_form()
```

File: core/includes/ajax.inc [hunk 2]
```diff
@@ -781,19 +790,33 @@ function ajax_pre_render_element($element) {
...
+    // substantially different for a JavaScript triggered submission. One such
+    // substantial difference is form elements that use #ajax['callback'] for
+    // determining which part of the form needs re-rendering. For that, we have
+    // a special 'system/ajax' route.
+    $path = isset($settings['callback']) ? 'system/ajax' : NULL;
+    $options = array();
+
+    $enabled_callbacks = variable_get('drupal_ajax_no_form_cache_enabled_callbacks', array());
+    if (isset($element['#ajax']['callback']) && isset($enabled_callbacks[$element['#ajax']['callback']]) && !isset($element['#ajax']['path'])) {
+      $query_parameters = drupal_get_query_parameters($_GET, array('q'));
```
Comment (klonos) [inline core/includes/ajax.inc:808]:
```suggestion
      $query_parameters = backdrop_get_query_parameters($_GET, array('q'));
```

File: core/includes/ajax.inc [hunk 3]
```diff
@@ -781,19 +790,33 @@ function ajax_pre_render_element($element) {
...
+    // negotiation takes care of formatting the response appropriately. However,
+    // 'path' and 'options' may be set when wanting server processing to be
+    // substantially different for a JavaScript triggered submission. One such
+    // substantial difference is form elements that use #ajax['callback'] for
+    // determining which part of the form needs re-rendering. For that, we have
+    // a special 'system/ajax' route.
+    $path = isset($settings['callback']) ? 'system/ajax' : NULL;
+    $options = array();
+
+    $enabled_callbacks = variable_get('drupal_ajax_no_form_cache_enabled_callbacks', array());
```
Comment (klonos) [inline core/includes/ajax.inc:806]: What's the equivalent config for this in Backdrop?

File: core/includes/form.inc [hunk 4]
```diff
@@ -128,7 +128,36 @@ function backdrop_get_form($form_id) {
...
-  return backdrop_build_form($form_id, $form_state);
+  // Disable redirection if the request was made via AJAX.
+  $ajax_post = isset($_POST['ajax_page_state']) && $_POST['form_id'] === $form_id;
+  if ($ajax_post) {
+    $form_state['no_redirect'] = TRUE;
+    $form_state['rebuild_info']['copy']['#build_id'] = TRUE;
+    $form_state['rebuild_info']['copy']['#action'] = TRUE;
+  }
+
+  $form = drupal_build_form($form_id, $form_state);
```
Comment (klonos) [inline core/includes/form.inc:139]:
```suggestion
  $form = backdrop_build_form($form_id, $form_state);
```

File: core/modules/simpletest/tests/ajax.test [hunk 5]
```diff
@@ -546,6 +546,63 @@ class AJAXFormPageCacheTestCase extends AJAXTestCase {
...
+/**
+ * Test Ajax forms when form cache is disabled for the ajax form.
+ */
+class AJAXFormNoCacheTestCase extends AJAXTestCase {
+  protected $profile = 'testing';
+
+  public function setUp() {
+    parent::setUp();
+
+    variable_set('backdrop_ajax_no_form_cache_enabled_callbacks', array(
```
Comment (klonos) [inline core/modules/simpletest/tests/ajax.test:558]: ...again, Backdrop config equivalent?

File: core/modules/simpletest/tests/ajax.test [hunk 6]
```diff
@@ -546,6 +546,63 @@ class AJAXFormPageCacheTestCase extends AJAXTestCase {
...
+    $expected = array(
+      'command' => 'updateBuildId',
+      'old' => $build_id_initial,
+      'new' => $build_id_first_ajax,
+    );
+    $this->assertCommand($commands, $expected, 'Build id change command issued on first AJAX submission');
+    $this->assertFalse($this->getFormFromCache($build_id_initial), 'AJAX form was not retrieved from cache_form.');
+
+    // Now disable all callbacks, allowing the ajax form to be cached.
+    variable_set('backdrop_ajax_no_form_cache_enabled_callbacks', array());
```
Comment (klonos) [inline core/modules/simpletest/tests/ajax.test:600]: ...same

**Comment by herbdool** [comment]
Restarting

**Comment by klonos** [comment]
...should be closed in favor of https://github.com/backdrop/backdrop/pull/2918

**Comment by quicksketch** [comment]
Yeah agreed, I looked at the approach the Drupal folks are taking and while it adds useful functionality it doesn't fix some of the big problems with the form cache that we have (both in D7 and Backdrop). I'm working on fixing up the tests.

**Pull request closed without merging**